File: 2024/minimum_add_to_make_parentheses_valid.py
#     Python
# Time complexity - O (N)
# Space complexity - O (1)

def minAddToMakeValid(self, s: str) -> int:
    open_parentheses = 0
    min_add = 0

    for char in s:
        if char == '(':
            open_parentheses += 1
        elif char == ')':
            if open_parentheses == 0:
                min_add += 1
            else:
                open_parentheses -= 1

    return min_add + open_parentheses

File: 2024/MinimumAddToMakeParenthesesValid.java
//            Java
//     Time complexity - O (N)
//     Space complexity - O (1)
public class MinimumAddToMakeParenthesesValid {

    public int minAddToMakeValid(String s) {
        int openParentheses = 0;
        int minAdd = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                openParentheses++;
            } else if (s.charAt(i) == ')') {
                if (openParentheses == 0) {
                    minAdd++;
                } else {
                    openParentheses--;
                }
            }
        }
        return minAdd + openParentheses;
    }
}
